controllers/admin/AdminBlockListingController.php

```diff
@@ -0,0 +1,205 @@
+    }
+
+    /**
+     * ajaxProcessSaveBlockContent
+     *
+     * @return bool|string
+     */
+    public function ajaxProcessSaveBlockContent()
+    {
+        $errors = array();
+
+        $picto = Tools::getValue('picto');
+        $id_block = Tools::getValue('id_block');
+        $psr_languages = (array) json_decode(Tools::getValue('lang_values'));
+        $type_link = Tools::getValue('typelink');
+        $id_cms = Tools::getValue('id_cms');
+        $newValues = array();
+
+        if (!empty($errors) && count($errors)) {
+            echo json_encode(array(
+                'errors' => $errors,
+            ));
+            return;
+        }
+        //$id_block++;
+        $blockPsr = new ReassuranceActivity($id_block);
+        $languages = Language::getLanguages();
+
+        foreach ($psr_languages as $key => $value) {
+            $newValues[$key]['title'] = $value->title;
+            $newValues[$key]['description'] = $value->description;
+            $newValues[$key]['url'] = $value->url;
+        }
+
+        foreach ($languages as $language) {
+            $blockPsr->title[$language['id_lang']] = $newValues[$language['id_lang']]['title'];
+            $blockPsr->description[$language['id_lang']] =  $newValues[$language['id_lang']]['description'];
+            if ($type_link == 2) {
+                $blockPsr->link[$language['id_lang']] =  $newValues[$language['id_lang']]['url'];
+            } else {
+                $blockPsr->link[$language['id_lang']] = '';
+            }
+        }
+
+        if (isset($id_cms) && $type_link == 1) {
+            $blockPsr->id_cms = $id_cms;
+            $link = Context::getContext()->link;
+            foreach ($languages as $language) {
+                $blockPsr->link[$language['id_lang']] = $link->getCMSLink($id_cms, null, null, $language['id_lang']);
+            }
+        }
+
+        if ($type_link == 'undefined') {
+            $type_link = 0;
+        }
+            
+
+        $blockPsr->icone = $picto;
+        if (empty($picto)) {
+            $blockPsr->icone_perso = '';
+        }
+        $blockPsr->type_link = $type_link;
+        
+        $blockPsr->date_add = date("Y-m-d H:i:s");
+        $blockPsr->date_update = date("Y-m-d H:i:s");
+
+        
+        if (isset($_FILES) && !empty($_FILES)) {
+            $picto_perso = $_FILES['file'];
+            $fileTmpName = $picto_perso['tmp_name'];
+            $filename = $picto_perso['name'];
+
+            // validateUpload return false if no error (false -> OK)
+            if (!ImageManager::validateUpload($picto_perso)) {
+                move_uploaded_file($fileTmpName, $this->module->folder_file_upload.$filename);
+                $blockPsr->icone_perso = $this->module->img_path_perso.'/'.$filename;
+                $blockPsr->icone = '';
+            } else {
+                $errors[] = ImageManager::validateUpload($picto_perso);
+                $this->ajaxDie(json_encode($errors));
```

Comment:
See, here you are handling the error case 👍 now I want you to do it everywhere 😄